File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/shadcn-command-item-state-variant-requires-value.ts
import type { SymbolDescriptor } from "../../semantic/scope-analysis.js";
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { findJsxAttribute } from "../../utils/find-jsx-attribute.js";
import { getImportedName } from "../../utils/get-imported-name.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { isTypeOnlyImport } from "../../utils/is-type-only-import.js";
import { resolveConstIdentifierAlias } from "../../utils/resolve-const-identifier-alias.js";
import { resolveShadcnUiComponentName } from "../../utils/resolve-shadcn-ui-component-name.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { splitTailwindClassName } from "../../utils/split-tailwind-class-name.js";
import { walkAst } from "../../utils/walk-ast.js";

const CMDK_MODULE = "cmdk";
const CMDK_MODULE_PATTERN = /^cmdk$/;
const COMMAND_MODULE_PATTERN = /(?:^|\/)ui\/(?:.*\/)?command$|^\.\.?\/(?:.*\/)?command$/;

// Matches a presence-only Tailwind data variant for the two boolean states —
// `data-[selected]:` / `data-[disabled]:` — anywhere in a token's variant
// chain. cmdk renders those attributes on every item as `"true"` OR
// `"false"`, so a presence selector matches both values and the style
// applies to every item unconditionally.
const PRESENCE_STATE_VARIANT_PATTERN = /(?:^|:)data-\[(selected|disabled)\](?=:)/;

const isCmdkValueImport = (symbol: SymbolDescriptor): boolean => {
  if (symbol.kind !== "import") return false;
  const declaration = symbol.declarationNode.parent;
  return Boolean(
    declaration &&
    isNodeOfType(declaration, "ImportDeclaration") &&
    !isTypeOnlyImport(declaration) &&
    declaration.source.value === CMDK_MODULE,
  );
};

// cmdk item spellings beyond what the shared resolver covers (`<CommandItem>`
// named import, `<Cmdk.CommandItem>` namespace member): `<Command.Item>` off
// the named `Command` object import and `<Cmdk.Command.Item>` off a
// namespace import.
const isCmdkItemElement = (elementName: EsTreeNode, context: RuleContext): boolean => {
  if (resolveShadcnUiComponentName(elementName, CMDK_MODULE_PATTERN, context) === "CommandItem") {
    return true;
  }
  if (!isNodeOfType(elementName, "JSXMemberExpression") || elementName.property.name !== "Item") {
    return false;
  }
  if (isNodeOfType(elementName.object, "JSXIdentifier")) {
    const symbol = resolveConstIdentifierAlias(elementName.object, context.scopes);
    return Boolean(
      symbol && isCmdkValueImport(symbol) && getImportedName(symbol.declarationNode) === "Command",
    );
  }
  if (
    !isNodeOfType(elementName.object, "JSXMemberExpression") ||
    !isNodeOfType(elementName.object.object, "JSXIdentifier") ||
    elementName.object.property.name !== "Command"
  ) {
    return false;
  }
  const namespaceSymbol = resolveConstIdentifierAlias(elementName.object.object, context.scopes);
  return Boolean(
    namespaceSymbol &&
    isCmdkValueImport(namespaceSymbol) &&
    isNodeOfType(namespaceSymbol.declarationNode, "ImportNamespaceSpecifier"),
  );
};

const isCommandItemElement = (elementName: EsTreeNode, context: RuleContext): boolean =>
  isCmdkItemElement(elementName, context) ||
  resolveShadcnUiComponentName(elementName, COMMAND_MODULE_PATTERN, context) === "CommandItem";

interface PresenceVariantUse {
  literalNode: EsTreeNode;
  state: string;
  token: string;
}

const findPresenceVariantUse = (
  text: string,
  literalNode: EsTreeNode,
): PresenceVariantUse | null => {
  for (const token of splitTailwindClassName(text)) {
    const match = PRESENCE_STATE_VARIANT_PATTERN.exec(token);
    if (match?.[1]) return { literalNode, state: match[1], token };
  }
  return null;
};

// Collects every statically-written class string reachable from the
// className attribute — a plain string, or string literals and template
// chunks anywhere inside the expression (`cn("…", className)`, cva config,
// clsx object keys, conditional branches).
const findPresenceVariantUseInClassName = (
  attribute: EsTreeNodeOfType<"JSXAttribute">,
): PresenceVariantUse | null => {
  const value = attribute.value;
  if (!value) return null;
  if (isNodeOfType(value, "Literal") && typeof value.value === "string") {
    return findPresenceVariantUse(value.value, value);
  }
  let use: PresenceVariantUse | null = null;
  walkAst(value, (node) => {
    if (use) return false;
    if (isNodeOfType(node, "Literal") && typeof node.value === "string") {
      use = findPresenceVariantUse(node.value, node);
    } else if (isNodeOfType(node, "TemplateLiteral")) {
      for (const quasi of node.quasis) {
        use ??= findPresenceVariantUse(quasi.value.cooked ?? quasi.value.raw ?? "", node);
      }
    }
  });
  return use;
};

export const shadcnCommandItemStateVariantRequiresValue = defineRule({
  id: "shadcn-command-item-state-variant-requires-value",
  title: "Command item styled by presence-only state variant",
  severity: "warn",
  category: "Correctness",
  requires: ["shadcn"],
  matchByOccurrence: true,
  recommendation:
    'Style command items with the value-aware variants data-[selected=true]: and data-[disabled=true]:, because cmdk renders both attributes on every item as "true" or "false".',
  create: (context: RuleContext) => ({
    JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
      if (!isCommandItemElement(node.name, context)) return;
      const classNameAttribute = findJsxAttribute(node.attributes, "className");
      if (!classNameAttribute) return;
      const use = findPresenceVariantUseInClassName(classNameAttribute);
      if (!use) return;
      context.report({
        node: use.literalNode,
        message: `cmdk renders data-${use.state} on every command item as "true" or "false", so "${use.token}" matches both values and styles every item. Use data-[${use.state}=true]: instead.`,
      });
    },
  }),
});

File: native/oxlint/rules/shadcn-command-item-state-variant-requires-value.rs
use oxc_ast::{
    AstKind,
    ast::{JSXAttributeValue, JSXElementName},
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::{GetSpan, Span};

use crate::{AstNode, context::LintContext, rule::Rule};

#[derive(Debug, Default, Clone)]
pub struct ShadcnCommandItemStateVariantRequiresValue;

declare_oxc_lint!(
    /// Require value-aware state variants on shadcn command items.
    ShadcnCommandItemStateVariantRequiresValue,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Require value-aware state variants on shadcn command items.",
);

impl Rule for ShadcnCommandItemStateVariantRequiresValue {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        if !has_capability_or_unspecified(ctx, "shadcn") {
            return;
        }
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        if !is_command_item_element(&opening_element.name, ctx) {
            return;
        }
        let Some(class_name_attribute) = find_jsx_attribute(opening_element, "className") else {
            return;
        };
        let Some((literal_span, state, token)) =
            find_presence_variant_use_in_class_name(class_name_attribute, ctx)
        else {
            return;
        };
        ctx.diagnostic(
            OxcDiagnostic::warn(format!(
                "cmdk renders data-{state} on every command item as \"true\" or \"false\", so \"{token}\" matches both values and styles every item. Use data-[{state}=true]: instead."
            ))
            .with_label(literal_span),
        );
    }
}

fn is_command_item_element<'a>(
    element_name: &JSXElementName<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    if resolve_shadcn_component_name(element_name, "command", ctx).as_deref()
        == Some("CommandItem")
    {
        return true;
    }
    resolve_jsx_import_api_path(element_name, |module_source| module_source == "cmdk", ctx)
        .is_some_and(|api_path| {
            matches!(
                api_path.as_slice(),
                [component_name] if component_name == "CommandItem"
            ) || matches!(
                api_path.as_slice(),
                [command_name, item_name] if command_name == "Command" && item_name == "Item"
            )
        })
}

fn find_presence_variant_use_in_class_name<'a>(
    attribute: &oxc_ast::ast::JSXAttribute<'a>,
    ctx: &LintContext<'a>,
) -> Option<(Span, &'static str, String)> {
    let value = attribute.value.as_ref()?;
    if let JSXAttributeValue::StringLiteral(string_literal) = value {
        return find_presence_variant_use(string_literal.value.as_str())
            .map(|(state, token)| (string_literal.span, state, token));
    }
    let JSXAttributeValue::ExpressionContainer(container) = value else {
        return None;
    };
    let expression = container.expression.as_expression()?;
    let expression_span = expression.span();
    for candidate in ctx.nodes().iter() {
        if !expression_span.contains_inclusive(candidate.span()) {
            continue;
        }
        match candidate.kind() {
            AstKind::StringLiteral(string_literal) => {
                if let Some((state, token)) =
                    find_presence_variant_use(string_literal.value.as_str())
                {
                    return Some((string_literal.span, state, token));
                }
            }
            AstKind::TemplateLiteral(template_literal) => {
                for quasi in &template_literal.quasis {
                    let text = quasi
                        .value
                        .cooked
                        .as_ref()
                        .map_or(quasi.value.raw.as_str(), |cooked| cooked.as_str());
                    if let Some((state, token)) = find_presence_variant_use(text) {
                        return Some((template_literal.span, state, token));
                    }
                }
            }
            _ => {}
        }
    }
    None
}

fn find_presence_variant_use(text: &str) -> Option<(&'static str, String)> {
    for token in tailwind_class_name_tokens(text) {
        let Some(state) = token.variants.iter().find_map(|variant| match *variant {
            "data-[selected]" => Some("selected"),
            "data-[disabled]" => Some("disabled"),
            _ => None,
        }) else {
            continue;
        };
        return Some((state, token.raw_token.to_string()));
    }
    None
}
